pkg/providers/aws/provider_postgres.go

```diff
@@ -0,0 +1,487 @@
+	if err != nil {
+		return nil, "error creating postgres instance", err
+	}
+	return nil, "", nil
+}
+
+func (p *AWSPostgresProvider) DeletePostgres(ctx context.Context, r *v1alpha1.Postgres) (v1alpha1.StatusMessage, error) {
+	// resolve postgres information for postgres created by provider
+	postgresCreateConfig, postgresDeleteConfig, stratCfg, err := p.getPostgresConfig(ctx, r)
+	if err != nil {
+		msg := "failed to retrieve aws postgres config"
+		return v1alpha1.StatusMessage(msg), err
+	}
+
+	// get provider aws creds so the postgres instance can be deleted
+	providerCreds, err := p.CredentialManager.ReconcileProviderCredentials(ctx, r.Namespace)
+	if err != nil {
+		msg := "failed to reconcile aws provider credentials"
+		return v1alpha1.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+
+	// setup aws postgres instance sdk session
+	instanceSvc := createPostgresService(stratCfg, providerCreds)
+
+	return p.deletePostgresInstance(ctx, r, instanceSvc, postgresCreateConfig, postgresDeleteConfig)
+}
+
+func (p *AWSPostgresProvider) deletePostgresInstance(ctx context.Context, pg *v1alpha1.Postgres, instanceSvc rdsiface.RDSAPI, postgresCreateConfig *rds.CreateDBInstanceInput, postgresDeleteConfig *rds.DeleteDBInstanceInput) (v1alpha1.StatusMessage, error) {
+	// the aws access key can sometimes still not be registered in aws on first try, so loop
+	pgs, err := getPostgresInstances(instanceSvc)
+	if err != nil {
+		return "error getting aws rds instances", err
+	}
+
+	// check and verify delete config
+	if err := p.verifyPostgresDeleteConfig(ctx, pg, postgresCreateConfig, postgresDeleteConfig); err != nil {
+		msg := "failed to verify aws postgres cluster configuration"
+		return v1alpha1.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+
+	// check if the instance has already been deleted
+	var foundInstance *rds.DBInstance
+	for _, i := range pgs {
+		if *i.DBInstanceIdentifier == *postgresDeleteConfig.DBInstanceIdentifier {
+			foundInstance = i
+			break
+		}
+	}
+
+	// check if instance does not exist, delete finalizer and credential secret
+	if foundInstance == nil {
+		// delete credential secret
+		p.Logger.Info("Deleting postgres secret")
+		sec := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      pg.Name + defaultCredSecSuffix,
+				Namespace: pg.Namespace,
+			},
+		}
+		err = p.Client.Delete(ctx, sec)
+		if err != nil && !k8serr.IsNotFound(err) {
+			msg := "failed to deleted postgres secrets"
+			return v1alpha1.StatusMessage(msg), errorUtil.Wrap(err, msg)
+		}
+
+		resources.RemoveFinalizer(&pg.ObjectMeta, DefaultFinalizer)
+		if err := p.Client.Update(ctx, pg); err != nil {
+			msg := "failed to update instance as part of finalizer reconcile"
+			return v1alpha1.StatusMessage(msg), errorUtil.Wrapf(err, msg)
+		}
+		return "", nil
+	}
+
+	// check if instance db exists and is available
+	if *foundInstance.DBInstanceStatus == "available" {
+		if *foundInstance.DeletionProtection {
```

Comment:
Why do we dereference here?

Reply:
Yeah we need to dereference these as they are aws variables, they use pointers.